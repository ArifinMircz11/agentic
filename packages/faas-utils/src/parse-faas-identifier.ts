import type { ParsedFaasIdentifier } from './types'
import { parseFaasUri } from './parse-faas-uri'

export function parseFaasIdentifier(
  identifier: string,
  { namespace }: { namespace?: string } = {}
): ParsedFaasIdentifier | undefined {
  if (!identifier) {
    return
  }

  let uri = identifier
  try {
    const { pathname } = new URL(identifier)
    uri = pathname
  } catch {}

  if (uri.startsWith('/')) {
    uri = uri.slice(1)
  }

  if (uri.endsWith('/')) {
    uri = uri.slice(0, -1)
  }

  if (!uri.length) {
    return
  }

  const hasNamespacePrefix = /^([a-zA-Z0-9-]{1,64}\/)/.test(uri)

  if (!hasNamespacePrefix) {
    if (namespace) {
      // add inferred namespace prefix (defaults to authenticated user's username)
      uri = `${namespace}/${uri}`
    } else {
      // throw new Error(`FaaS identifier is missing namespace prefix or you must be authenticated [${uri}]`)
      return
    }
  }

  return parseFaasUri(uri)
}
