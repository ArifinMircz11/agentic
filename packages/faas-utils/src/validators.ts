import emailValidator from 'email-validator'
import isRelativeUrl from 'is-relative-url'

export const usernameRe = /^[a-zA-Z0-9-]{1,64}$/
export const passwordRe = /^.{3,1024}$/

export const projectNameRe = /^[a-z0-9-]{3,64}$/
export const deploymentHashRe = /^[a-z0-9]{8}$/

export const projectRe = /^[a-zA-Z0-9-]{1,64}\/[a-z0-9-]{3,64}$/
export const deploymentRe = /^[a-zA-Z0-9-]{1,64}\/[a-z0-9-]{3,64}@[a-z0-9]{8}$/

// service names may be any valid JavaScript identifier
// TODO: should service names be any label?
export const serviceNameRe = /^[a-zA-Z_][a-zA-Z0-9_]*$/
export const servicePathRe = /^\/[a-zA-Z0-9\-._~%!$&'()*+,;=:/]*$/

export function email(value: string): boolean {
  return emailValidator.validate(value)
}

export function username(value: string): boolean {
  return !!value && usernameRe.test(value)
}

export function password(value: string): boolean {
  return !!value && passwordRe.test(value)
}

export function projectName(value: string): boolean {
  return !!value && projectNameRe.test(value)
}

export function deploymentHash(value: string): boolean {
  return !!value && deploymentHashRe.test(value)
}

export function project(value: string): boolean {
  return !!value && projectRe.test(value)
}

export function deployment(value: string): boolean {
  return !!value && deploymentRe.test(value)
}

export function serviceName(value: string): boolean {
  return !!value && serviceNameRe.test(value)
}

export function servicePath(value: string): boolean {
  return !!value && servicePathRe.test(value) && isRelativeUrl(value)
}
