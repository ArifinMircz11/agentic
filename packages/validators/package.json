{
  "name": "@agentic/platform-validators",
  "version": "0.1.0",
  "description": "Validation utils for the Agentic platform.",
  "author": "Travis Fischer <travis@transitivebullsh.it>",
  "license": "UNLICENSED",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/transitive-bullshit/agentic-platform.git",
    "directory": "packages/validators"
  },
  "type": "module",
  "source": "./src/index.ts",
  "types": "./src/index.ts",
  "sideEffects": false,
  "exports": {
    ".": "./src/index.ts"
  },
  "scripts": {
    "test": "run-s test:*",
    "test:lint": "eslint .",
    "test:typecheck": "tsc --noEmit",
    "test:unit": "vitest run"
  },
  "dependencies": {
    "@paralleldrive/cuid2": "^2.2.2",
    "email-validator": "^2.0.4",
    "is-relative-url": "^4.0.0"
  }
}
