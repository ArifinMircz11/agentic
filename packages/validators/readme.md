<p align="center">
  <a href="https://agentic.so/publishing">
    <img alt="Agentic" src="/apps/web/public/agentic-publishing-social-image-dark-github.jpg" width="640">
  </a>
</p>

<p>
  <a href="https://github.com/transitive-bullshit/agentic/actions/workflows/main.yml"><img alt="Build Status" src="https://github.com/transitive-bullshit/agentic/actions/workflows/main.yml/badge.svg" /></a>
  <a href="https://www.npmjs.com/package/@agentic/platform-validators"><img alt="NPM" src="https://img.shields.io/npm/v/@agentic/platform-validators.svg" /></a>
  <a href="https://prettier.io"><img alt="Prettier Code Formatting" src="https://img.shields.io/badge/code_style-prettier-brightgreen.svg" /></a>
</p>

# @agentic/platform-validators <!-- omit from toc -->

> Core schemas and validators shared across the Agentic platform.

- [Website](https://agentic.so/publishing)
- [Docs](https://docs.agentic.so)

> [!TIP]
> You likely don't need this package directly. See [@agentic/cli](https://github.com/transitive-bullshit/agentic/tree/main/packages/cli), [@agentic/platform](https://github.com/transitive-bullshit/agentic/tree/main/packages/platform), and [@agentic/platform-tool-client](https://github.com/transitive-bullshit/agentic/tree/main/packages/platform-tool-client) for more public-facing packages.

## Install

```bash
npm i @agentic/platform-validators
```

## Usage

```ts
import { parseProjectIdentifier } from '@agentic/platform-validators'

const parsedProjectIdentifier = parseProjectIdentifier('@agentic/search')
```

## Identifiers

### Project Identifier

- `@username/project-slug`
- `@team-slug/project-slug`

**Examples:**

- `@agentic/search`

### Deployment Identifier

- `${projectIdentifier}` will implicitly use `${projectIdentifier}@latest`
- `${projectIdentifier}@latest` (the most recently published deployment)
- `${projectIdentifier}@dev` (the most recently pushed deployment)
- `${projectIdentifier}@deploymentHash` (a specific deployment)
- `${projectIdentifier}@version` (a specific published deployment specified via `semver`)

**Examples:**

- `@agentic/search`
- `@agentic/search@latest`
- `@agentic/search@1.0.0`

### Tool Identifier

- `${deploymentIdentifier}/tool_name`

**Examples:**

- `@agentic/search/search`
- `@agentic/search@latest/search`
- `@agentic/search@1.0.0/search`

### Tool Names

- Must start with a letter or underscore
- Can include only letters, numbers, and underscores
- Use either camelCase or snake_case consistently across all tools

[OpenAI vs Anthropic vs Google vs MCP tool name restrictions](https://chatgpt.com/share/68419382-73a0-8007-afce-0ded7d9f05e7)

## License

[GNU AGPL 3.0](https://choosealicense.com/licenses/agpl-3.0/)
