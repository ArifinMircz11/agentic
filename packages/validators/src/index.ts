export * from './parse-faas-identifier'
export * from './parse-faas-uri'
export type * from './types'
export * as validators from './validators'
