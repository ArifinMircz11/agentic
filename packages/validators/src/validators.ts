import { isCuid } from '@paralleldrive/cuid2'
import emailValidator from 'email-validator'

import type { ParseIdentifierOptions } from './types'
import { namespaceBlacklist } from './namespace-blacklist'
import { parseDeploymentIdentifier } from './parse-deployment-identifier'
import { parseProjectIdentifier } from './parse-project-identifier'

export const namespaceRe = /^[a-z0-9-]{1,256}$/
export const passwordRe = /^.{3,1024}$/

export const projectNameRe = /^[a-z0-9-]{1,256}$/
export const deploymentHashRe = /^[a-z0-9]{8}$/

export const toolNameRe = /^[a-zA-Z_][a-zA-Z0-9_]{0,63}$/

export function isValidEmail(value: string): boolean {
  return emailValidator.validate(value)
}

export function isValidNamespace(value?: string): boolean {
  return !!value && namespaceRe.test(value)
}

export function isNamespaceAllowed(value?: string): boolean {
  return !!value && isValidNamespace(value) && !namespaceBlacklist.has(value)
}

export function isValidUsername(value?: string): boolean {
  return isValidNamespace(value)
}

export function isValidTeamSlug(value?: string): boolean {
  return isValidNamespace(value)
}

export function isValidPassword(value?: string): boolean {
  return !!value && passwordRe.test(value)
}

export function isValidProjectName(value?: string): boolean {
  return !!value && projectNameRe.test(value)
}

export function isValidDeploymentHash(value?: string): boolean {
  return !!value && deploymentHashRe.test(value)
}

export function isValidProjectIdentifier(
  value?: string,
  opts?: ParseIdentifierOptions
): boolean {
  return !!parseProjectIdentifier(value, opts)
}

export function isValidDeploymentIdentifier(
  value?: string,
  opts?: ParseIdentifierOptions
): boolean {
  return !!parseDeploymentIdentifier(value, opts)
}

export function isValidToolName(value?: string): boolean {
  return !!value && toolNameRe.test(value)
}

export function isValidCuid(value?: string): boolean {
  return !!value && isCuid(value)
}
