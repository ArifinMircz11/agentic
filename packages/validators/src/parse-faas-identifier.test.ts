import { expect, test } from 'vitest'

import { parseFaasIdentifier } from './parse-faas-identifier'
import * as validators from './validators'

function success(...args: Parameters<typeof parseFaasIdentifier>) {
  const result = parseFaasIdentifier(...args)
  expect(result).toBeTruthy()
  expect(result!.projectIdentifier).toBeTruthy()
  expect(result!.version || result!.deploymentHash).toBeTruthy()
  expect(validators.projectIdentifier(result!.projectIdentifier)).toBe(true)
  expect(validators.toolPath(result!.toolPath)).toBe(true)

  if (result!.deploymentHash) {
    expect(validators.deploymentHash(result!.deploymentHash)).toBe(true)
    expect(validators.deploymentIdentifier(result!.deploymentIdentifier!)).toBe(
      true
    )
  }

  expect(result).toMatchSnapshot()
}

function error(...args: Parameters<typeof parseFaasIdentifier>) {
  const result = parseFaasIdentifier(...args)
  expect(result).toBeUndefined()
}

test('URL prefix success', () => {
  success('https://api.saasify.sh/username/foo-bar@01234567/foo')
  success('/username/foo-bar@01234567/foo')
  success('https://api.saasify.sh/username/foo-bar@01234567/foo/bar/456/123')
  success('/username/foo-bar@01234567/foo/bar/456/123')
  success('/username/foo-bar@01234567/foo/bar/456/123')
  success('/username/foo-bar@latest/foo/bar/456/123')
  success('/username/foo-bar@dev/foo/bar/456/123')
  success('/username/foo-bar@2.1.0/foo/bar/456/123')
})

test('URL prefix error', () => {
  error('https://api.saasify.sh/2/proxy/username/foo-bar@01234567/foo')
  error('/call/username/foo-bar@01234567/foo')
  error('//username/foo-bar@01234567/foo')
})

test('URL suffix success', () => {
  success('username/foo-bar@01234567/foo/')
  success('username/foo-bar@latest/foo/')
  success('username/foo-bar@dev/foo/')
  success('username/foo-bar@2.1.0/foo/')
})

test('URL suffix error', () => {
  error('username/foo-bar@01234567/fooðŸ˜€')
  error('username/Foo-Bar@dev/foo/')
})

test('URL prefix and suffix success', () => {
  success('https://api.saasify.sh/username/foo-bar@01234567/foo/')
  success('https://api.saasify.sh/username/foo-bar@01234567/foo/bar/123')
})

test('namespace success', () => {
  success('https://api.saasify.sh/foo-bar@01234567/foo', {
    namespace: 'username'
  })
  success('/foo-bar@01234567/foo', { namespace: 'username' })
  success('/foo-bar@01234567/foo', { namespace: 'username' })
  success('/foo-bar@01234567/foo/', { namespace: 'username' })
  success('https://api.saasify.sh/foo-bar@01234567/foo/bar/123', {
    namespace: 'username'
  })
  success('/foo-bar@01234567/foo/bar/123', { namespace: 'username' })
  success('/foo-bar@latest/foo/bar/123', { namespace: 'username' })
  success('/foo-bar@dev/foo/bar/123', { namespace: 'username' })
  success('/foo-bar@1.2.3/foo/bar/123', { namespace: 'username' })
})

test('namespace error', () => {
  error('https://api.saasify.sh/foo-bar@01234567/foo')
  error('https://api.saasify.sh/foo-bar@latest/foo')
  error('/foo-bar@01234567/foo')
  error('/foo-bar@dev/foo')
  error('/foo-bar@01234567/foo')
  error('/foo-bar@01234567/foo/')
  error('/foo-bar@01234567/foo/bar/123')
  error('/foo-bar@0latest/foo/bar/123')
})
