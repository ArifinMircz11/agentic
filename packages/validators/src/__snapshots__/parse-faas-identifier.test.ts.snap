// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`URL prefix and suffix success 1`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`URL prefix and suffix success 2`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/123",
}
`;

exports[`URL prefix success 1`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`URL prefix success 2`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`URL prefix success 3`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/456/123",
}
`;

exports[`URL prefix success 4`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/456/123",
}
`;

exports[`URL prefix success 5`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/456/123",
}
`;

exports[`URL prefix success 6`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/456/123",
  "version": "latest",
}
`;

exports[`URL prefix success 7`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/456/123",
  "version": "dev",
}
`;

exports[`URL prefix success 8`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/456/123",
  "version": "2.1.0",
}
`;

exports[`URL suffix success 1`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`URL suffix success 2`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
  "version": "latest",
}
`;

exports[`URL suffix success 3`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
  "version": "dev",
}
`;

exports[`URL suffix success 4`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
  "version": "2.1.0",
}
`;

exports[`namespace success 1`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`namespace success 2`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`namespace success 3`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`namespace success 4`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`namespace success 5`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/123",
}
`;

exports[`namespace success 6`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/123",
}
`;

exports[`namespace success 7`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/123",
  "version": "latest",
}
`;

exports[`namespace success 8`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/123",
  "version": "dev",
}
`;

exports[`namespace success 9`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo/bar/123",
  "version": "1.2.3",
}
`;
